### DISCUSSION TITLE
Support for custom operators for mongo, mysql, postgres and snowflake

### DISCUSSION THREAD

**Discussion by viramjainfyno** [react-querybuilder/react-querybuilder#718] [General]
**Description of the feature**

If any table in the database includes a date field, I've developed a custom operator to create relative queries (e.g., last 7 days) and store them. These queries ensure that whenever I execute them, I get updated data.

I've successfully converted the rules to database queries and stored them. However, I'm encountering difficulties in extracting the original rules to render them on the frontend.

Below is the code for the custom operator:
```js
  const customFormatQuery = (query, options) => {
    const processRules = rules => {
      return rules.map(rule => {
        if (typeof rule === 'string') {
          // If it's a logical operator, return it as is.
          return rule
        }

        // Process nested rules by creating a new object with updated rules.
        if (rule?.rules) {
          return {
            ...rule,
            rules: processRules(rule.rules)
          }
        }

        // Handle 'lastXDays' operator.
        if (rule?.operator === 'lastXDays') {
          options.valueProcessor = (r, options) => {
            const found = fields?.find(f => f?.name === r?.field || f?.label === r?.field)
            if (!found) return defaultValueProcessorByRule(r, options)

            if (
              typeof r?.value === 'string' &&
              (r?.value?.startsWith('NOW() - INTERVAL') ||
                r?.value?.startsWith('current_date - interval') ||
                r?.value?.startsWith(//`dateadd('days'//`))
            ) {
              return r?.value
            } else if (typeof r?.value === 'object') {
              return JSON.stringify(r?.value)
            }

            return defaultValueProcessorByRule(r, options)
          }

          const days = parseInt(rule.value, 10)
          const dateRule = {
            id: '1',
            field: rule?.field || 'NA',
            operator: '>=',
            valueSource: 'value',
            value: rule?.value
          }

          if (isMysql) {
            dateRule.value = /`NOW() - INTERVAL ${days} DAY/`
          } else if (isPostgresql) {
            dateRule.value = /`current_date - interval '${days} days'/`
          } else if (isSnowFlake) {
            dateRule.value = /`dateadd('days', -${days}, current_date)/`
          } else if (isMongoDB) {
            dateRule.value = {
              $expr: {
                $lte: [
                  {
                    $dateDiff: {
                      startDate: /`$${dateRule.field}/`,
                      endDate: '$$NOW',
                      unit: 'day'
                    }
                  },
                  days
                ]
              }
            }
          }

          return dateRule
        }

        return rule
      })
    }

    const newQuery = {
      ...query,
      rules: processRules(query?.rules || [])
    }

    return formatQuery(newQuery, options)
  }

  let filter = customFormatQuery(filterReactQuery, options)

  const operators = defaultOperators?.filter(operator => !operator.name.includes('doesNot'))

  operators.push({ name: 'lastXDays', label: 'last X days' })
```

**Comment by viramjainfyno** [comment]
This is how I receive the filter in the API response:

```json
{
  "$expr": {
    "$lte": [
      {
        "$dateDiff": {
          "startDate": "$date",
          "endDate": "$$NOW",
          "unit": "day"
        }
      },
      7
    ]
  }
}
```

**Comment by jakeboone02** [comment]
Questions:

- Are you using [`parseMongoDB`](https://react-querybuilder.js.org/docs/utils/import#mongodb) with [`additionalOperators`](https://react-querybuilder.js.org/docs/utils/import#custom-operators) to convert the API response to the RQB object type?
- What does your value editor code look like? We implemented a similar "relative date range" feature where I work and I'm curious what your UX is.

Recommendations:

- I would _strongly_ recommend storing queries in the RQB native format on the server, not in whatever database flavor. That way you don't have to convert back and forth just to display and update the queries in the UI. You can always generate the executable queries with `formatQuery` as and when they need to be run against the database.
- Use [`transformQuery`](https://react-querybuilder.js.org/docs/utils/misc#transformquery) instead of that `customFormatQuery` function. It's effectively the same, but it has some extra safeguards and features like immutability. (It doesn't call `formatQuery` at the end so you'd have to do that separately.) You might be able to use your rule processor (the one that starts on this line: `return rules.map(rules => ...`) as is. You'd end up with something like this:
  ```ts
  formatQuery(
    transformQuery(filterReactQuery, { ruleProcessor: yourRuleProcessorHere })
  )
  ```

**Reply by viramjainfyno** [reply]
Hello

I am currently using the 6.5.0 version of the library which does not have support for additionalOperators. Also, in the latest version, additionalOperators is only available for parseMongoDB and not for parseSQL (I have a requirement to support both the database types)

Attaching a screenshot of the UX for your reference:

<img width="1377" alt="image" src="https://github.com/react-querybuilder/react-querybuilder/assets/150130308/13b18129-9d13-4bec-abc1-2a09676c7fab">

**Reply by jakeboone02** [reply]
Is there anything blocking you from upgrading to v7?

Again, I would strongly recommend storing and updating the queries purely with the RQB object format, and converting to SQL or MongoDB only when necessary, if at all possible. But if you need to parse MongoDB and SQL syntax for some reason, just know that: 1) `parseMongoDB` from v7 will produce RQB query objects that are compatible with v6 and earlier, and 2) `parseSQL` is pretty much a black box and only supports a subset of standard SQL syntax/operators. I won't rule it out completely because I haven't looked at it in a long time, but I doubt there will ever be a similar feature to "additionalOperators" for `parseSQL`.

**Reply by jakeboone02** [reply]
By the way, I like the styling on your query builder, especially the "remove" icon middle-aligned on the right border of the group.

**Reply by viramjainfyno** [reply]
Thank you for your response and appreciation.

I was able to accomplish what I required by implementing what you had suggested (storing and updating the queries purely with the RQB object format, and converting to SQL or MongoDB only when necessary)

Thank you for your time.
